src/Functions/generateUUIDv7.cpp

```diff
@@ -0,0 +1,351 @@
+        {
+            uint64_t counter = uuid[6] & 0x0f;
+            counter = (counter << 8) | uuid[7];
+            counter = (counter << 6) | (uuid[8] & 0x3f);
+            counter = (counter << 8) | uuid[9];
+            counter = (counter << 8) | uuid[10];
+            counter = (counter << 8) | uuid[11];
+            return counter;
+        }
+
+        void generate(UUID & newUUID)
+        {
+            uint64_t timestamp = 0;
+            /// Get timestamp of the previous uuid
+            for (int i = 0; i != 6; ++i)
+                timestamp = (timestamp << 8) | uuid[i];
+
+            const uint64_t unix_time_ms = getTimestampMs();
+            // continue incrementing counter when clock slightly goes back or when counter overflow happened during the previous UUID generation
+            bool need_to_increment_counter = (timestamp == unix_time_ms || timestamp < unix_time_ms + 10000);
+            uint64_t counter = 0;
+            if (need_to_increment_counter)
+                counter = getCounter();
+            else
+                timestamp = unix_time_ms;
+
+            bool counter_incremented = false;
+            if (need_to_increment_counter)
+            {
+                if (++counter == counter_limit)
+                {
+                    ++timestamp;
+                    // counter bytes will be filled by the random data
+                }
+                else
+                {
+                    uuid[6] = counter >> 38;
+                    uuid[7] = counter >> 30;
+                    uuid[8] = counter >> 24;
+                    uuid[9] = counter >> 16;
+                    uuid[10] = counter >> 8;
+                    uuid[11] = counter;
+                    counter_incremented = true;
+                }
+            }
+
+            fillTimestamp(uuid, timestamp);
+
+            // Get the required number of random bytes: 4 in the case of incrementing existing counter, 10 in the case of renewing counter
+            memcpy(
+                uuid.data() + (counter_incremented ? next_count_random_data_offset : random_data_offset),
+                &newUUID,
+                counter_incremented ? next_count_random_data_count : random_data_count);
+
+            store(newUUID, uuid);
+        }
+    };
+
+    struct ThreadLocalCounter
+    {
+        static constexpr auto name = "generateUUIDv7WithFastCounter";
+        struct Data : CounterDataCommon
+        {
+            // Implement counter monotony only within one thread so function doesn't require mutexes and doesn't affect performance of the same function running simultenaously on other threads
+            static inline thread_local UUIDAsArray uuid_data;
+
+            Data()
+                : CounterDataCommon(uuid_data)
+            {}
+        };
+    };
+
+    struct GlobalCounter
+    {
+        static constexpr auto name = "generateUUIDv7WithCounter";
+        struct Data : std::lock_guard<std::mutex>, CounterDataCommon
```

Comment:
There is too much inheritance here and inheriting from standard library classes is bad practice. Could you try composition instead?

(if not possible, then we can go with the current code)

Reply:
Will go with composition.